crates/gitbutler-branch-actions/src/remote.rs

```diff
@@ -60,32 +62,50 @@ pub struct RemoteCommit {
     pub conflicted: bool,
 }
 
-/// Return information on all local branches, while skipping gitbutler-specific branches in `refs/heads`.
-///
-/// Note to be confused with `list_branches()`, which is used for the new branch listing.
+/// Finds all branches matching a given name, which can be at most one local branch,
+/// and any number of branches (on different remotes).
 ///
 /// # Previous notes
 /// For legacy purposes, this is still named "remote" branches, but it's actually
 /// a list of all the normal (non-gitbutler) git branches.
-pub fn list_local_branches(ctx: &CommandContext) -> Result<Vec<RemoteBranch>> {
+pub fn find_git_branches(ctx: &CommandContext, branch_name: &str) -> Result<Vec<RemoteBranch>> {
     let default_target = default_target(&ctx.project().gb_dir())?;
-
-    let mut remote_branches = vec![];
-    let remotes = ctx.repository().remotes()?;
-    for (branch, _) in ctx
-        .repository()
-        .branches(None)
-        .context("failed to list remote branches")?
-        .flatten()
+    let repo = ctx.repository();
+    let mut branches = vec![];
+    let remotes_raw = repo.remotes()?;
+    let remotes = remotes_raw.iter().flatten().collect_vec();
+
+    let mut all_branches: Vec<_> = remotes
+        .iter()
+        .filter_map(|remote| {
+            let branch_name = format!("{}/{}", remote, branch_name);
+            repo.find_branch(&branch_name, BranchType::Remote)
+                .ok()
+                .and_then(|branch| branch_to_remote_branch(&branch, &[remote]).ok())
+                .unwrap_or(None)
+        })
+        .collect();
+
+    if let Some(local_branch) = repo
+        .find_branch(branch_name, BranchType::Local)
+        .ok()
+        .and_then(|branch| {
+            branch_to_remote_branch(&branch, &remotes)
+                .map_err(|err| {
+                    tracing::warn!(?err, "Ignoring branch");
+                })
+                .ok()
+        })
+        .unwrap_or(None)
     {
-        let branch = match branch_to_remote_branch(&branch, &remotes) {
-            Ok(Some(b)) => b,
-            Ok(None) => continue,
-            Err(err) => {
-                tracing::warn!(?err, "Ignoring branch");
-                continue;
-            }
-        };
+        all_branches.push(local_branch);
+    }
```

Comment:
Could we write this one as
```rs
// Let the compiler know we're swallowing the error
let _ = repo
         .find_branch(branch_name, BranchType::Local)
         .and_then(|branch| {
             let branch = branch_to_remote_branch(&branch, &remotes)?;
             all_branches.push(branch);
             Ok(())
         });
```

Reply:
The question mark does not work inside of `.and_then(..)`. Also, `branch_to_remote_branch` returns an `Option`, but we could perhaps change that to an error.

Reply:
The question mark does work inside of `and_then`. If you've got a function or closure whose return type is `Option`, then you can use `?` on any functions which return `Option` inside of it, and similar for `Result`.


I don't see any particularly good reason why branch_to_remote_branch should be `Result<Option<>>` as it a failure to convert. If it returns just a `Result` then my snippet should work.